src/main/java/net/fabricmc/loader/impl/ObjectShareImpl.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright 2016 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.loader.impl;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.function.BiConsumer;
+
+import net.fabricmc.loader.api.ObjectShare;
+
+final class ObjectShareImpl implements ObjectShare {
+	private final Map<String, Object> values = new HashMap<>();
+	private final Map<String, List<BiConsumer<String, Object>>> pending = new HashMap<>();
+
+	@Override
+	public synchronized Object get(String key) {
+		validateKey(key);
+
+		return values.get(key);
+	}
+
+	@Override
+	public synchronized Object put(String key, Object value) {
+		validateKey(key);
+		Objects.requireNonNull(value, "null value");
+
+		Object prev = values.put(key, value);
+
+		if (prev == null) {
```

Comment:
Shouldn't this refire any callbacks if it is replacing the object? But, see my comment on whenAvailable()